source/slang/slang-ir-simplify-cfg.cpp

```diff
@@ -0,0 +1,74 @@
+#include "slang-ir-simplify-cfg.h"
+
+#include "slang-ir-insts.h"
+#include "slang-ir.h"
+
+namespace Slang
+{
+
+void processFunc(IRFunc* func)
+{
+    auto firstBlock = func->getFirstBlock();
+    if (!firstBlock)
+        return;
+
+    List<IRBlock*> workList;
+    HashSet<IRBlock*> processedBlock;
+    workList.add(func->getFirstBlock());
+    while (workList.getCount())
+    {
+        auto block = workList.getFirst();
+        workList.fastRemoveAt(0);
+        while (block)
+        {
+            // If `block` does not end with an unconditional branch, bail.
+            if (block->getTerminator()->getOp() != kIROp_unconditionalBranch)
+                break;
+            auto branch = as<IRUnconditionalBranch>(block->getTerminator());
+            auto successor = branch->getTargetBlock();
+            // Only perform the merge if `block` is the only predecessor of `successor`.
+            if (successor->hasMoreThanOneUse())
+                break;
```

Comment:
I don't know if we *need* to note this in the comment, but it is important that the main case where `successor->hasMoreThanOneUse()` could be true even when `successor` might have only one predecessor is if `successor` is the "merge" point for some control-flow construct. E.g., if you have:

```
while( true )
{
    if(A())
    {
        B();
        break;
    }
    C();
}
D();
```

then the block with `D()` is the only successor of the block with `B()`, but we wouldn't be able to merge the two as long as the `while` loop remains in the IR, because the instruction for that loop references the `D()` block as its merge point. It is *required* that we don't perform a merge in a case like the above, because as written `B()` and `D()` will typically execute with different "active sets" of threads, and thus with different wave masks. Any merge of those two blocks could affect the value returned by wave-mask-dependent operations under `B()` and/or `D()`.

Reply:
Yeah, and this is exactly one of the bugs I have fixed when implementing this.